File: step1/src/avoidance.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"
#include "sensor_msgs/LaserScan.h"
#include "geometry_msgs/Twist.h"

float angY, angZ, linX;

void laserScanCallback(const sensor_msgs::LaserScan laser)
{
    bool bol_str8=true;
    
    //right detected, turn left
    for (const auto range : {laser.ranges[0], laser.ranges[1], laser.ranges[2], laser.ranges[3], laser.ranges[4],})
    {
    if ( range<3 && range>1 )
    {
        ROS_INFO("TURN LEFT");
        linX = 0.0f;
        angY = 1.0f;
        angZ = -1.0f;
        bol_str8=false;
    }
    }

    //left detected, turn right
    for (const auto range : {laser.ranges[5], laser.ranges[6], laser.ranges[7], laser.ranges[8], laser.ranges[9],})
    {
      if( range<3 && range>1 )
      {
        ROS_INFO("TURN RIGHT");
        linX = 0.0f;
        angY = 1.0f;
        angZ = 1.0f;
        bol_str8=false;
      }
    }

    //nothing detected, go straight
    if (bol_str8==true)
    {
        ROS_INFO("GO STRAIGHT");
        linX = 0.3f;
        angY = 0.0f;
        angZ = 0.0f;
    }

}
  

int main(int argc, char **argv)
{
  ros::init(argc, argv, "hLaserReader");
  ros::NodeHandle nodeHandler;
  ros::Subscriber hokuyoSubscriber = nodeHandler.subscribe("/m2wr/laser/scan", 1, laserScanCallback);
  ros::Publisher pubVelocity = nodeHandler.advertise<geometry_msgs::Twist>("/cmd_vel", 10);

  ros::Rate rate(10);
  linX = 0.0f;
  angY = 0.0f;
  angZ = 0.0f;
  while (ros::ok()) {
    geometry_msgs::Twist msg;
    msg.linear.x = linX;
    msg.angular.y = angY;
    msg.angular.z = angZ;

    pubVelocity.publish(msg);
    rate.sleep();
    ros::spinOnce();
  }

  ros::spin();

  return 0;
}

File: step1/src/avoidance.py
#!/usr/bin/python3

import rospy
import time
from std_msgs.msg import String
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Twist

def laserScanCallback(laser):
    global angY, angZ, linX

    if (0.2 < laser.ranges[2] < 0.5) and (0.2 < laser.ranges[1] < 0.7) and (0.2 < laser.ranges[0] < 0.5):
        rospy.loginfo("TURN RIGHT")
        linX = 0.0
        angZ = 1.0
    elif (0.2 < laser.ranges[2] < 0.5) and (0.2 < laser.ranges[1] < 0.7) and not (0.2 < laser.ranges[0] < 0.5):
        rospy.loginfo("TURN LEFT")
        linX = 0.0
        angZ = -1.0
    elif (0.2 < laser.ranges[2] < 0.5) and not (0.2 < laser.ranges[1] < 0.7):
        rospy.loginfo("GO STRAIGHT")
        linX = 0.3
        angY = 0.0
        angZ = 0.0
    elif not (0.2 < laser.ranges[2] < 0.5) and (0.2 < laser.ranges[1] < 0.7):
        rospy.loginfo("TURN RIGHT")
        linX = 0.0
        angZ = 1.0
    elif not (0.2 < laser.ranges[2] < 0.5) and not (0.2 < laser.ranges[1] < 0.7) and (0.2 < laser.ranges[0] < 0.5):
        rospy.loginfo("TURN RIGHT")
        linX = 0.0
        angZ = 1.0
    elif not (0.2 < laser.ranges[2] < 0.5) and not (0.2 < laser.ranges[1] < 0.7) and not (0.2 < laser.ranges[0] < 0.5):
        rospy.loginfo("GO STRAIGHT")
        linX = 0.3
        angY = 0.0
        angZ = 0.0
    elif (0.2 < laser.ranges[2] < 1) and (0.2 < laser.ranges[1] < 0.7) and (0.2 < laser.ranges[0] < 1):
        rospy.loginfo("TURN TURN")
        linX = 0.0
        angZ = -1.0
        time.sleep(1)

    # Publish velocity
    msg = Twist()
    msg.linear.x = linX
    msg.angular.y = angY
    msg.angular.z = angZ
    pubVelocity.publish(msg)

if __name__ == "__main__":
    rospy.init_node("hLaserReader")
    pubVelocity = rospy.Publisher("/cmd_vel", Twist, queue_size=10)
    rospy.Subscriber("/m2wr/laser/scan", LaserScan, laserScanCallback)
    rate = rospy.Rate(10)
    linX = 0.0
    angY = 0.0
    angZ = 0.0
    while not rospy.is_shutdown():
        rate.sleep()

    rospy.spin()